rholang/src/main/k/rholang.k

```diff
@@ -1,132 +1,1039 @@
 module RHO-SYNTAX
 imports SUBSTITUTION
 
-// Syntax for processes
-syntax Proc ::=
-              // Empty process
-                "Nil"
+// We need processes, including variables
+syntax Proc ::= ProcNoVars
+              | Id
+
+// We need processes, excluding variables
+syntax ProcNoVars ::=
+              // Ground terms, including the empty process and expressions
+                Ground
               // Listen
-              | "for(" Name "<-" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<-" Name ")" "{" Proc "}" [binder]
               // Persistent listen
-              | "for(" Name "<=" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<=" Name ")" "{" Proc "}" [binder]
               | "contract" Name "(" Name ")" "=" "{" Proc "}" [binder]
               // Send
-              | Name "!(" Proc ")"
+              | Name "!(" Procs ")"
               // Evaluate
               | "*" Name
               // New
-              | "new" Id "in" "{" Proc "}" [binder]
+              | "new" Ids "in" "{" Proc "}" [binder]
               // Parallel
               | Proc "|" Proc [left]
-              // Misc
+
+// We put "Nil" and Exp in the same syntactic category, since they are
+// structurally equivalent as ground properties
+syntax Ground ::= "Nil"
               | Exp
 
-// Syntax for Names; we allow variables to take the place of names, but note
-// that no rholang program will compile if there are free variables left in
-// the program. Thus the variable must be bound to a name of the form @Proc
-syntax Name ::= "@" Proc
+// We need names with no variables
+// (We need put the Proc through some algorithm which puts it into its normal form; will come later)
+syntax NameNoVars ::= "@" Proc
```

Comment:
Why isn't this `"@" ProcNoVars`?

Reply:
Sometimes we need a name of the form @x, as long as x is only locally free, for example:
`for( @ {@x!(Nil)} <- @"chan"){ @"chan"!(x) } | @"chan"!( @"chan2"!(Nil) )`

Reply:
If we do `"@" ProcNoVars`, we can't have correct syntax of that form.

Reply:
I trust you, but don't understand yet. I think I might have misunderstood your comments. _I'll have to look at what "putting the Proc into the normal form" means_...

Reply:
All I meant by the comments above was that if we have `"@" ProcNoVars`, the syntax `@x` would not be valid for a name, but it's important that it be valid for a name because of examples like the one I gave.

Ah, the "normal form" isn't implemented. That's a bit of a red herring comment, sorry. The normal form is just a way of having only one unique way to represent two equivalent processes/names. For example, the names `@{10+2}` and `@{5 + 7}` are the same as `@12`, since `10+2 = 5+7 = 12`. Also `P | Q = Q | P` so `@P | Q = @Q | P`. Ideally we'd like to be able to tell that if I'm listening on `@P | Q` and sending on `@Q | P`, I'm going to be working on the same channel. That's an issue that we have to deal with at some point (but it's not here yet).